Add test cases from the Allen (1983) paper

genelkim opened the pull request:
I tried writing a test case, but it keeps failing to compile. And due to my limited recent experience with Haskell, I'm having trouble figuring out what I'm doing wrong. Adam, could you take a look and let me know how I might fix this test case? Below is the error message in case that helps.

```
Build profile: -w ghc-8.6.5 -O1
In order, the following will be built (use -v for more details):
 - allen-0.1.0.0 (test:allen-test) (file test/Spec.hs changed)
Preprocessing test suite 'allen-test' for allen-0.1.0.0..
Building test suite 'allen-test' for allen-0.1.0.0..
[2 of 2] Compiling Main             ( test/Spec.hs, /mnt/c/Users/genel/code/allen/dist-newstyle/build/x86_64-linux/ghc-8.6.5/allen-0.1.0.0/t/allen-test/build/allen-test/allen-test-tmp/Main.o )

test/Spec.hs:115:31: error:
    • Couldn't match type ‘GHC.Word.Word16’
                     with ‘transformers-0.5.6.2:Control.Monad.Trans.State.Lazy.StateT
                             IntervalGraph Data.Functor.Identity.Identity RelationBits’
      Expected type: transformers-0.5.6.2:Control.Monad.Trans.State.Lazy.StateT
                       IntervalGraph Data.Functor.Identity.Identity RelationBits
        Actual type: RelationBits
    • In a stmt of a 'do' block:
        srRelationBits <- bitsFromString "pmMP"
      In the expression:
        do r <- interval
           s <- interval
           l <- interval
           srRelationBits <- bitsFromString "pmMP"
           ....
      In an equation for ‘calc’:
          calc
            = do r <- interval
                 s <- interval
                 l <- interval
                 ....
    |
115 |             srRelationBits <- bitsFromString "pmMP"
    |                               ^^^^^^^^^^^^^^^^^^^^^

test/Spec.hs:116:31: error:
    • Couldn't match type ‘GHC.Word.Word16’
                     with ‘transformers-0.5.6.2:Control.Monad.Trans.State.Lazy.StateT
                             IntervalGraph Data.Functor.Identity.Identity RelationBits’
      Expected type: transformers-0.5.6.2:Control.Monad.Trans.State.Lazy.StateT
                       IntervalGraph Data.Functor.Identity.Identity RelationBits
        Actual type: RelationBits
    • In a stmt of a 'do' block: slRelationBits <- bitsFromString "om"
      In the expression:
        do r <- interval
           s <- interval

           l <- interval
           srRelationBits <- bitsFromString "pmMP"
           ....
      In an equation for ‘calc’:
          calc
            = do r <- interval
                 s <- interval
                 l <- interval
                 ....
    |
116 |             slRelationBits <- bitsFromString "om"
    |                               ^^^^^^^^^^^^^^^^^^^

test/Spec.hs:123:29: error:
    • Couldn't match type ‘GHC.Word.Word16’
                     with ‘transformers-0.5.6.2:Control.Monad.Trans.State.Lazy.StateT
                             IntervalGraph Data.Functor.Identity.Identity RelationBits’
      Expected type: transformers-0.5.6.2:Control.Monad.Trans.State.Lazy.StateT
                       IntervalGraph Data.Functor.Identity.Identity RelationBits
        Actual type: RelationBits
    • In a stmt of a 'do' block:
        expectedBits <- bitsFromString "pPomDsSFe"
      In the expression:
        do r <- interval
           s <- interval
           l <- interval
           srRelationBits <- bitsFromString "pmMP"
           ....
      In an equation for ‘calc’:
          calc
            = do r <- interval
                 s <- interval
                 l <- interval
                 ....
    |
123 |             expectedBits <- bitsFromString "pPomDsSFe"
    |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Archaversine reviewed (changes requested):
For the most part everything looks fine. When you're using `bitsFromString`, you should use `let` instead of `<-`. So for example, these lines:

```haskell
srRelationBits <- bitsFromString "pmMP"
slRelationBits <- bitsFromString "om"
```

would become:

```haskell
let srRelationBits = bitsFromString "pmMP"
    slRelationBits = bitsFromString "om"
```

and that should fix your compilation errors (`expectedBits` would also need to be changed). `<-` is used for binding monadic values (in this case anything wrapped in `Allen`) and `let` is used for typical variable assignments. Since `bitsFromString` returns a `Bool` instead of an `Allen Bool`, there's no need to bind it's value. Hopefully this explanation makes sense, let me know if you have any more questions!

genelkim commented:
Thanks! The remaining questions I have seem to be the lack of some fundamental Haskell knowledge. I'll need to spend a bit of time refreshing myself on how exactly Haskell works to understand the nuances of monadic values vs. general variables.

genelkim commented:
I fixed that and added a few more test cases from the paper. They're failing right now. This isn't really a problem since that just means we need to finish implementing something.

However, I'd like you to look over the test cases and make sure that I'm using the library as intended. I don't quite have a grasp on when network constraints are propagated (or how we trigger the constraint propagation manually, if that's how it's done) so I may be pulling out inferred constraints in the wrong places. I think it would also be good for you to walk through how exactly the codebase works for me since I'm still a bit confused (at least partially due to my limited Haskell understanding).

Archaversine commented:
I just looked over the test cases, everything you did looks correct. The benefit of using a monad to model these calculations is that we can abstract away all interaction with the actual graph of intervals. So for example when you do `x <- interval`, behind the scenes you are creating a brand new interval, adding it to the graph, and setting the value of `x` to the interval's ID in the network (this was implemented in this way to play nicely with Haskell's immutability). This has not been implemented yet, but when you `assume` a set of relations between two intervals, the graph will automatically update in the background to propgate the relations to all the intervals. If you want to manually interact with the graph, you can use the [Control.Monad.State functions](https://hackage.haskell.org/package/mtl-2.3.1/docs/Control-Monad-State-Lazy.html). I will add more detail about this in the README.md and add more examples to make it more intuitive.
